javascript/ql/lib/semmle/javascript/frameworks/ExpressValidator.qll

```diff
@@ -0,0 +1,114 @@
+/**
+ * Models of npm modules that are used with Express servers.
+ */
+
+import javascript
+private import semmle.javascript.frameworks.HTTP
+private import semmle.javascript.frameworks.Express
+private import semmle.javascript.security.dataflow.Xss
+
+/**
+ * Models of the express-validator npm module.
+ */
+module ExpressValidator {
+  /**
+   * The middleware instance that is used to validate the request.
+   */
+  class MiddlewareInstance extends DataFlow::InvokeNode {
+    private string validator;
+
+    MiddlewareInstance() {
+      exists(DataFlow::SourceNode middleware |
+        (
+          // query('search').notEmpty().escape()
+          middleware = DataFlow::moduleMember("express-validator", ["query", "param"]).getACall() and
+          validator = "parameter"
+          or
+          // body('search').notEmpty().escape()
+          middleware = DataFlow::moduleMember("express-validator", "body").getACall() and
+          validator = "body"
+          or
+          // cookie('search').notEmpty().escape()
+          middleware = DataFlow::moduleMember("express-validator", "cookie").getACall() and
+          validator = "cookie"
+          or
+          // header('search').notEmpty().escape()
+          middleware = DataFlow::moduleMember("express-validator", "header").getACall() and
+          validator = "header"
+        ) and
+        isSafe(middleware)
+      |
+        this = middleware
+      )
+    }
+
+    /**
+     * Gets the name iof the parameter that is safe.
+     */
+    string getSafeParameterName() { this.getArgument(0).mayHaveStringValue(result) }
+
+    /**
+     * Gets the type of the validator (parameter, body, etc)
+     */
+    string getValidatorType() { result = validator }
+
+    /**
+     * Gets the route handler that is validated.
+     */
+    Express::RouteHandler getRouteHandler() {
+      exists(Express::RouteSetup route |
+        this.getAstNode().getParent*() = route.getARouteHandlerNode().getAstNode()
+      |
+        result = route.getLastRouteHandlerNode()
+      )
```

Comment:
There's an opportunity to use the routing trees library here. If you add
```codeql
private import semmle.javascript.Routing
```
You should be able to just do this: (but please test first)
```suggestion
    Routing::getRouteHandler(result).isGuardedBy(this)
```
This should also cover cases where the middleware is installed in a `.use()` call in front of a multiple routes or subrouters.